### ISSUE TITLE
Infinite loop fetching repos since refactoring PR#24

### ISSUE THREAD

**Issue by mauro-justgiving** [CircleCI-Public/CircleCI-Env-Inspector#35]
The project was running fine last week.
Since PR#24 `getCircleCIRepos()` is being called in an infinite loop by `getPaginatedData()` on the first of 5 accounts/organisations I work on.
In case it's relevant, I'm not an administrator of that organisation.

Right now I cannot find the documentation for CircleCI's /api/private but I'll try and debug the issue myself.

I've added a single debug line:

```
diff --git a/src/utils/utils.mts b/src/utils/utils.mts
index 740f027..9e48b5e 100644
--- a/src/utils/utils.mts
+++ b/src/utils/utils.mts
@@ -54,6 +54,7 @@ export async function getPaginatedData<T>(
       identifier,
       pageToken
     );
+    console.log(`responseBody.items.length: ${responseBody.items.length}, pageToken: ****${pageToken.slice(-4)}`);
     if (response.ok && responseBody.items.length > 0)
       items.push(...responseBody.items);
     pageToken = responseBody.next_page_token;
```

The output is:

```
Running Docker image...                                                                                                                
? Enter your CircleCI API token [hidden]
Found 5 accounts.
Fetching data for <withheld> from GitHub...  1/5
  Fetching Contexts...
responseBody.items.length: 1, pageToken: ****
responseBody.items.length: 0, pageToken: ****
  Fetching Projects...
responseBody.items.length: 49, pageToken: ****
responseBody.items.length: 49, pageToken: ****yMtw
responseBody.items.length: 50, pageToken: ****SpBA
responseBody.items.length: 49, pageToken: ****NGIA
responseBody.items.length: 49, pageToken: ****eYBw
responseBody.items.length: 49, pageToken: ****vswA
responseBody.items.length: 49, pageToken: ****1stw
responseBody.items.length: 50, pageToken: ****QcTw
responseBody.items.length: 49, pageToken: ****U-3A
responseBody.items.length: 49, pageToken: ****o22Q
responseBody.items.length: 49, pageToken: ****FwfI
responseBody.items.length: 49, pageToken: ****p4Qg
responseBody.items.length: 50, pageToken: ****ojsI
responseBody.items.length: 49, pageToken: ****bprA
responseBody.items.length: 49, pageToken: ****yAkA
responseBody.items.length: 49, pageToken: ****ih4Y
responseBody.items.length: 49, pageToken: ****QrGA
responseBody.items.length: 50, pageToken: ****b6LQ
responseBody.items.length: 49, pageToken: ****YreQ

```

**Comment by BeeeQueue** [comment]
Can confirm. Happening to me as an org admin

**Comment by KyleTryon** [comment]
Hello folks, if you have a moment, please try pulling the latest version and let us know if you are still experiencing this issue.

@mauro-justgiving `/api/private` is indeed not documented. We "stole" this from the front end application because it provided better data than what we found in the documented APIs.

**Comment by mauro-justgiving** [comment]
Hello @KyleTryon , I've pulled the latest version and unfortunately I see the same issue.
I can see in #41 you are suggesting an organisation administrator key, and I am not an administrator. Could it be what's causing my issue?
If there is a way I can help debugging let me know! 

```
Running Docker image...                                                                                                                
? Enter your CircleCI API token [hidden]
Found 5 accounts.
Fetching data for <omitted> from GitHub...  1/5
  Fetching Contexts...
responseBody.items.length: 1, pageToken: ****
responseBody.items.length: 0, pageToken: ****
  Fetching Projects...
responseBody.items.length: 49, pageToken: ****
responseBody.items.length: 49, pageToken: ****M9pw
responseBody.items.length: 50, pageToken: ****ye-g
responseBody.items.length: 49, pageToken: ****mBxg
responseBody.items.length: 49, pageToken: ****vT4A
responseBody.items.length: 49, pageToken: ****7DCU
responseBody.items.length: 49, pageToken: ****rtDw
responseBody.items.length: 50, pageToken: ****6Ec0
responseBody.items.length: 49, pageToken: ****CPgQ
responseBody.items.length: 49, pageToken: ****hNog
responseBody.items.length: 49, pageToken: ****DUk8
responseBody.items.length: 49, pageToken: ****dECQ
responseBody.items.length: 50, pageToken: ****ewtI
...
```

**Comment by BeeeQueue** [comment]
Like I said it still happens if you're an admin in the org.

**Comment by KyleTryon** [comment]
Your token should not be relevant. It's a bit hard to debug but I am looking into it. I can not reproduce the issue with several large organizations.

**Comment by mauro-justgiving** [comment]
Thanks. For context the organization I'm having trouble with has 335 projects in total, of which 180 unavailable, according to when I ran the project before the refactoring.

**Comment by BeeeQueue** [comment]
When the API endpoint reaches the last page it seems to just keep returning page tokens with the same repos it already returned earlier in, resulting in infinite duplicates.

Our org has 1200~ GH repos that CCI recognizes of which 300~ or so had ssh keys.

The most I saw before cancelling it was around 10k items fetched, with 9k duplicates

I ended up editing the source to stop after reaching about 1500 items and then removing duplicates before saving it

**Comment by KyleTryon** [comment]
Thank you for the added context @BeeeQueue . I am looking into this!

**Comment by mauro-justgiving** [comment]
I have put together a PR to introduce a circuit breaker. It allowed me to have the script complete.
@BeeeQueue would you like to give it a go?
@KyleTryon what do you think?

**Comment by mauro-justgiving** [comment]
Resolved with #44
